// @synergenius/flow-weaver/dist/api/index.js
/**
 * @module api
 *
 * # Flow Weaver API
 *
 * High-level API for working with Flow Weaver workflows.
 *
 * ## Parse → Validate → Generate Pipeline
 *
 * ```
 * Source File (.ts)
 *       ↓
 *   parseWorkflow()     → TWorkflowAST
 *       ↓
 *   validateWorkflow()  → ValidationResult
 *       ↓
 *   generateCode()      → Executable code
 * ```
 *
 * Or use `compileWorkflow()` for all steps in one call.
 *
 * ## Key Exports
 *
 * - **Parsing**: {@link parseWorkflow}
 * - **Validation**: {@link validateWorkflow}
 * - **Generation**: {@link generateCode}, {@link generateInPlace}
 * - **Compilation**: {@link compileWorkflow}
 * - **Query**: `getNode()`, `getNodeType()`, `findConnections()`
 * - **Manipulation**: `addNode()`, `removeNode()`, `addConnection()`, etc.
 * - **Builder**: `WorkflowBuilder` for programmatic AST construction
 */
export { compileWorkflow } from './compile.js';
export { generateCode } from './generate.js';
export { generateInPlace, hasInPlaceMarkers, stripGeneratedSections, MARKERS, } from './generate-in-place.js';
export { parseWorkflow } from './parse.js';
export { transformWorkflow } from './transform.js';
export { validateWorkflow } from './validate.js';
export * from './manipulation.js';
export { withValidation, withMinimalValidation, withoutValidation, validatePortReference, portReferencesEqual, formatPortReference, generateUniqueNodeId, assertNodeTypeExists, assertNodeExists, assertNodeNotExists, } from './helpers.js';
export * from './query.js';
export * from './builder.js';
export * from './workflow-file-operations.js';
export * from './templates.js';
export * from './patterns.js';
