// @synergenius/flow-weaver/dist/types/branded-ports.js
/**
 * Branded Port Types
 *
 * Provides compile-time type safety for port handling using TypeScript's
 * branded types pattern. Prevents accidental mixing of port categories.
 *
 * Categories:
 * - MandatoryPort: Reserved ports (execute, onSuccess, onFailure, start, success, failure)
 * - ScopedPort: Ports belonging to a scope (forEach, map, etc.)
 * - RegularPort: All other ports
 */
// =============================================================================
// Port Name Constants
// =============================================================================
/**
 * Reserved port name constants.
 * Use these instead of magic strings.
 */
export const PortName = {
    // External mandatory ports
    EXECUTE: 'execute',
    ON_SUCCESS: 'onSuccess',
    ON_FAILURE: 'onFailure',
    // Scoped mandatory ports
    START: 'start',
    SUCCESS: 'success',
    FAILURE: 'failure',
};
/**
 * External mandatory port names.
 */
export const EXTERNAL_MANDATORY_PORTS = [
    PortName.EXECUTE,
    PortName.ON_SUCCESS,
    PortName.ON_FAILURE,
];
/**
 * Scoped mandatory port names.
 */
export const SCOPED_MANDATORY_PORTS = [
    PortName.START,
    PortName.SUCCESS,
    PortName.FAILURE,
];
// =============================================================================
// Type Guards
// =============================================================================
/**
 * Type guard to check if a port is mandatory.
 * Works for both external (execute, onSuccess, onFailure) and scoped (start, success, failure).
 */
export function isMandatoryPort(name, port) {
    const isScoped = port.scope !== undefined;
    if (isScoped) {
        return SCOPED_MANDATORY_PORTS.includes(name);
    }
    return EXTERNAL_MANDATORY_PORTS.includes(name);
}
/**
 * Type guard to check if a port is scoped.
 */
export function isScopedPort(port) {
    return port !== undefined && typeof port.scope === 'string';
}
/**
 * Type guard to check if a port is regular (not mandatory).
 */
export function isRegularPort(name, port) {
    return !isMandatoryPort(name, port);
}
// =============================================================================
// Assertion Functions
// =============================================================================
/**
 * Assert that a port is mandatory and return typed version.
 * @throws Error if port is not mandatory
 */
export function toMandatoryPort(name, port) {
    if (!isMandatoryPort(name, port)) {
        throw new Error(`"${name}" is not a mandatory port`);
    }
    return { name, port: port };
}
/**
 * Assert that a port is scoped and return typed version.
 * @throws Error if port is not scoped
 */
export function toScopedPort(port) {
    if (!isScopedPort(port)) {
        throw new Error(`Port is not a scoped port (no scope property)`);
    }
    return port;
}
/**
 * Get the order value from a port's metadata.
 * Returns Infinity if no order is set.
 */
export function getPortOrder(port) {
    const order = port.metadata?.order;
    return typeof order === 'number' ? order : Infinity;
}
/**
 * Create an ordered port list from a port record.
 * Sorts by metadata.order, with unordered ports at the end.
 */
export function createOrderedPorts(ports) {
    const entries = Object.entries(ports).map(([name, port]) => ({
        name,
        port,
        order: getPortOrder(port),
    }));
    // Sort by order (lower first), then by name for stable sort
    entries.sort((a, b) => {
        if (a.order !== b.order) {
            return a.order - b.order;
        }
        return a.name.localeCompare(b.name);
    });
    return entries;
}
