// @synergenius/flow-weaver/dist/testing/mock-llm.js
/**
 * Mock LLM Provider for deterministic agent workflow testing.
 *
 * Usage in vitest/jest tests:
 * ```typescript
 * import { createMockLlmProvider } from 'flow-weaver/testing';
 *
 * const mockLlm = createMockLlmProvider([
 *   { match: /search/, response: { content: null, toolCalls: [{ id: '1', name: 'search', arguments: { q: 'test' } }], finishReason: 'tool_calls' } },
 *   { match: /./, response: { content: 'Default answer', toolCalls: [], finishReason: 'stop' } },
 * ]);
 *
 * // Inject globally for compiled workflows
 * (globalThis as unknown as { __fw_llm_provider__?: LLMProvider }).__fw_llm_provider__ = mockLlm;
 *
 * // After test
 * expect(mockLlm.getCallCount()).toBe(2);
 * expect(mockLlm.getTotalTokens()).toBe(150);
 * ```
 */
const DEFAULT_USAGE = {
    promptTokens: 50,
    completionTokens: 25,
    totalTokens: 75,
};
const FALLBACK_RESPONSE = {
    content: '[Mock: no matching response configured]',
    toolCalls: [],
    finishReason: 'stop',
    usage: DEFAULT_USAGE,
};
/**
 * Creates a mock LLM provider with deterministic response matching.
 *
 * Responses are matched in order against the last user message content.
 * First match wins. If no response matches, a fallback response is returned.
 *
 * @param responses - Array of match/response pairs, checked in order
 * @param options - Configuration options
 * @returns A mock provider with call recording and token tracking
 */
export function createMockLlmProvider(responses = [], options) {
    const calls = [];
    const useCounts = new Map();
    const defaultUsage = options?.defaultUsage ?? DEFAULT_USAGE;
    const fallback = {
        ...FALLBACK_RESPONSE,
        ...(options?.fallbackResponse ?? {}),
        usage: options?.fallbackResponse?.usage ?? defaultUsage,
    };
    function findMatch(lastMessage) {
        for (let i = 0; i < responses.length; i++) {
            const def = responses[i];
            // Check max uses
            if (def.maxUses !== undefined) {
                const used = useCounts.get(i) ?? 0;
                if (used >= def.maxUses)
                    continue;
            }
            const matches = typeof def.match === 'string'
                ? lastMessage.includes(def.match)
                : def.match.test(lastMessage);
            if (matches) {
                return { index: i, response: def };
            }
        }
        return null;
    }
    return {
        async chat(messages, chatOptions) {
            const lastUserMessage = [...messages].reverse().find((m) => m.role === 'user')?.content ?? '';
            const match = findMatch(lastUserMessage);
            let response;
            let matchedIndex;
            if (match) {
                matchedIndex = match.index;
                useCounts.set(matchedIndex, (useCounts.get(matchedIndex) ?? 0) + 1);
                response = {
                    content: match.response.response.content ?? null,
                    toolCalls: match.response.response.toolCalls ?? [],
                    finishReason: match.response.response.finishReason ?? 'stop',
                    usage: match.response.usage ?? defaultUsage,
                };
            }
            else {
                matchedIndex = -1;
                response = { ...fallback };
            }
            calls.push({
                messages: [...messages],
                options: chatOptions,
                matchedIndex,
                response,
                timestamp: Date.now(),
            });
            return response;
        },
        getCalls() {
            return [...calls];
        },
        getCallCount() {
            return calls.length;
        },
        getTotalTokens() {
            return calls.reduce((sum, call) => sum + (call.response.usage?.totalTokens ?? 0), 0);
        },
        getTokenUsage() {
            return calls.map((call) => call.response.usage ?? defaultUsage);
        },
        reset() {
            calls.length = 0;
            useCounts.clear();
        },
        getCallsForResponse(index) {
            return calls.filter((c) => c.matchedIndex === index);
        },
    };
}
