// @synergenius/flow-weaver/dist/testing/mock-approval.js
/**
 * Mock Approval Provider for deterministic human-approval workflow testing.
 *
 * Mirrors the mock LLM provider pattern — pattern matching on prompts,
 * call recording, and configurable responses.
 *
 * Usage in vitest/jest tests:
 * ```typescript
 * import { createMockApprovalProvider } from 'flow-weaver/testing';
 *
 * const mockApproval = createMockApprovalProvider([
 *   { match: /expense/, response: { approved: true, reviewer: 'alice' } },
 *   { match: /delete/, response: { approved: false, response: 'Too risky' } },
 *   { match: /./, response: { approved: true } },
 * ]);
 *
 * // Inject globally for compiled workflows
 * (globalThis as unknown as { __fw_approval_provider__?: ApprovalProvider })
 *   .__fw_approval_provider__ = mockApproval;
 *
 * // After test
 * expect(mockApproval.getCallCount()).toBe(2);
 * ```
 */
const FALLBACK_RESULT = {
    approved: false,
    response: '[Mock: no matching approval response configured]',
};
/**
 * Creates a mock approval provider with deterministic response matching.
 *
 * Responses are matched in order against the approval request prompt.
 * First match wins. If no response matches, a fallback result is returned.
 *
 * @param responses - Array of match/response pairs, checked in order
 * @param options - Configuration options
 * @returns A mock provider with call recording
 */
export function createMockApprovalProvider(responses = [], options) {
    const calls = [];
    const useCounts = new Map();
    const fallback = options?.fallbackResponse ?? FALLBACK_RESULT;
    function findMatch(prompt) {
        for (let i = 0; i < responses.length; i++) {
            const def = responses[i];
            // Check max uses
            if (def.maxUses !== undefined) {
                const used = useCounts.get(i) ?? 0;
                if (used >= def.maxUses)
                    continue;
            }
            const matches = typeof def.match === 'string'
                ? prompt.includes(def.match)
                : def.match.test(prompt);
            if (matches) {
                return { index: i, response: def };
            }
        }
        return null;
    }
    return {
        async requestApproval(request) {
            const match = findMatch(request.prompt);
            let result;
            let matchedIndex;
            if (match) {
                matchedIndex = match.index;
                useCounts.set(matchedIndex, (useCounts.get(matchedIndex) ?? 0) + 1);
                result = { ...match.response.response };
            }
            else {
                matchedIndex = -1;
                result = { ...fallback };
            }
            calls.push({
                request: { ...request },
                matchedIndex,
                result,
                timestamp: Date.now(),
            });
            return result;
        },
        getCalls() {
            return [...calls];
        },
        getCallCount() {
            return calls.length;
        },
        reset() {
            calls.length = 0;
            useCounts.clear();
        },
        getCallsForResponse(index) {
            return calls.filter((c) => c.matchedIndex === index);
        },
    };
}
