// @synergenius/flow-weaver/dist/testing/recorder.js
/**
 * Recording LLM Provider — wraps a real provider and records all interactions.
 *
 * Usage:
 * ```typescript
 * import { createRecordingProvider } from 'flow-weaver/testing';
 *
 * const recorder = createRecordingProvider(realLlmProvider);
 *
 * // Use recorder.provider as the LLM provider in your workflow
 * (globalThis as unknown as { __fw_llm_provider__?: LLMProvider }).__fw_llm_provider__ = recorder.provider;
 *
 * // Run workflow...
 *
 * // Get the recording for later replay
 * const recording = recorder.getRecording();
 * const json = JSON.stringify(recording, null, 2);
 * // Save to disk, fixture file, etc.
 * ```
 */
/**
 * Creates a recording LLM provider that wraps a real provider.
 *
 * Every call to `chat()` is passed through to the real provider,
 * and the input/output/timing is recorded for later replay.
 *
 * @param realProvider - The real LLM provider to wrap
 * @param metadata - Optional metadata to include in the recording
 * @returns A recording provider with access to the recorded data
 */
export function createRecordingProvider(realProvider, metadata) {
    const steps = [];
    const createdAt = new Date().toISOString();
    const provider = {
        async chat(messages, options) {
            const startTime = Date.now();
            const timestamp = new Date(startTime).toISOString();
            const response = await realProvider.chat(messages, options);
            const durationMs = Date.now() - startTime;
            steps.push({
                index: steps.length,
                input: {
                    messages: [...messages],
                    ...(options ? { options: { ...options } } : {}),
                },
                output: { ...response },
                timestamp,
                durationMs,
            });
            return response;
        },
    };
    return {
        provider,
        getRecording() {
            return {
                steps: steps.map((s) => ({ ...s })),
                metadata: { ...(metadata ?? {}) },
                createdAt,
            };
        },
        getCallCount() {
            return steps.length;
        },
        reset() {
            steps.length = 0;
        },
    };
}
