// @synergenius/flow-weaver/dist/deployment/index.js
/**
 * Flow Weaver Deployment System
 *
 * Unified abstractions for deploying workflows across CLI, HTTP, and serverless contexts.
 *
 * @module deployment
 */
// Core - Executor
export { UnifiedWorkflowExecutor, createExecutor } from './core/executor.js';
// Core - Adapters
export { CliRequestAdapter, HttpRequestAdapter, LambdaRequestAdapter, VercelRequestAdapter, CloudflareRequestAdapter, createAdapter, } from './core/adapters.js';
// Core - Formatters
export { formatCliResponse, formatHttpResponse, formatLambdaResponse, formatCloudflareResponse, formatError, } from './core/formatters.js';
export { DEFAULT_CONFIG, DEFAULT_SERVER_CONFIG, DEFAULT_EXECUTION_CONFIG, getDefaultConfig, } from './config/defaults.js';
export { loadConfig, loadConfigSync, getConfigValue } from './config/loader.js';
// OpenAPI
export { OpenAPIGenerator, generateOpenAPIJson, generateOpenAPIYaml, } from './openapi/generator.js';
export { SchemaConverter, schemaConverter, } from './openapi/schema-converter.js';
// Export Targets
export { BaseExportTarget, ExportTargetRegistry, } from './targets/base.js';
export { LambdaTarget } from './targets/lambda.js';
export { VercelTarget } from './targets/vercel.js';
export { CloudflareTarget } from './targets/cloudflare.js';
export { InngestTarget } from './targets/inngest.js';
// Convenience: Create a pre-configured target registry
import { ExportTargetRegistry } from './targets/base.js';
import { LambdaTarget } from './targets/lambda.js';
import { VercelTarget } from './targets/vercel.js';
import { CloudflareTarget } from './targets/cloudflare.js';
import { InngestTarget } from './targets/inngest.js';
/**
 * Default export target registry with all built-in targets
 */
export function createTargetRegistry() {
    const registry = new ExportTargetRegistry();
    registry.register(new LambdaTarget());
    registry.register(new VercelTarget());
    registry.register(new CloudflareTarget());
    registry.register(new InngestTarget());
    return registry;
}
/**
 * Get names of all supported export targets
 */
export function getSupportedTargetNames() {
    return ['lambda', 'vercel', 'cloudflare', 'inngest'];
}
