// @synergenius/flow-weaver/dist/deployment/core/executor.js
/**
 * Unified workflow executor
 *
 * Provides a single entry point for executing workflows across all contexts.
 */
import { randomUUID } from 'crypto';
import { executeWorkflowFromFile } from '../../mcp/workflow-executor.js';
/**
 * Unified workflow executor
 *
 * Provides consistent execution behavior across CLI, HTTP, and serverless contexts.
 */
export class UnifiedWorkflowExecutor {
    registry;
    defaultTimeout;
    production;
    constructor(options = {}) {
        this.registry = options.registry;
        this.defaultTimeout = options.defaultTimeout ?? 30000;
        this.production = options.production ?? false;
    }
    /**
     * Execute a workflow with the given request
     */
    async execute(request) {
        const startTime = Date.now();
        const requestId = request.context.requestId || randomUUID();
        try {
            // Resolve the workflow
            const workflow = await this.resolveWorkflow(request.workflowId);
            if (!workflow) {
                return this.createErrorResponse(request.workflowId, requestId, 'WORKFLOW_NOT_FOUND', `Workflow "${request.workflowId}" not found`, startTime);
            }
            // Execute with timeout if specified
            const timeout = request.context.timeout ?? this.defaultTimeout;
            const includeTrace = request.context.includeTrace && !this.production;
            const result = await this.executeWithTimeout(workflow.filePath, request.params, {
                workflowName: workflow.functionName,
                production: this.production || request.context.environment === 'production',
                includeTrace,
            }, timeout, request.abortSignal);
            return {
                success: true,
                workflowId: request.workflowId,
                result: result.result,
                executionTime: Date.now() - startTime,
                trace: includeTrace ? result.trace : undefined,
                requestId,
            };
        }
        catch (error) {
            // Check for specific error types
            if (error instanceof Error) {
                if (error.name === 'AbortError' || error.message.includes('abort')) {
                    return this.createErrorResponse(request.workflowId, requestId, 'CANCELLED', 'Workflow execution was cancelled', startTime);
                }
                if (error.message.includes('timeout') || error.message.includes('timed out')) {
                    return this.createErrorResponse(request.workflowId, requestId, 'TIMEOUT', `Workflow execution timed out after ${request.context.timeout ?? this.defaultTimeout}ms`, startTime);
                }
                // Check if it's a validation error
                if (error.message.includes('validation') || error.message.includes('invalid')) {
                    return this.createErrorResponse(request.workflowId, requestId, 'VALIDATION_ERROR', error.message, startTime, error.stack);
                }
                // General execution error
                return this.createErrorResponse(request.workflowId, requestId, 'EXECUTION_ERROR', error.message, startTime, error.stack);
            }
            // Unknown error
            return this.createErrorResponse(request.workflowId, requestId, 'INTERNAL_ERROR', String(error), startTime);
        }
    }
    /**
     * Execute a workflow directly from a file path
     */
    async executeFromFile(filePath, params, context = {}) {
        const fullContext = {
            source: context.source ?? 'cli',
            environment: context.environment ?? 'development',
            requestId: context.requestId ?? randomUUID(),
            includeTrace: context.includeTrace ?? true,
            timeout: context.timeout,
        };
        const request = {
            workflowId: filePath,
            params,
            context: fullContext,
        };
        return this.executeFileDirectly(filePath, request);
    }
    /**
     * Execute a file directly without registry lookup
     */
    async executeFileDirectly(filePath, request) {
        const startTime = Date.now();
        const requestId = request.context.requestId;
        try {
            const timeout = request.context.timeout ?? this.defaultTimeout;
            const includeTrace = request.context.includeTrace && !this.production;
            const result = await this.executeWithTimeout(filePath, request.params, {
                production: this.production || request.context.environment === 'production',
                includeTrace,
            }, timeout, request.abortSignal);
            return {
                success: true,
                workflowId: result.functionName,
                result: result.result,
                executionTime: Date.now() - startTime,
                trace: includeTrace ? result.trace : undefined,
                requestId,
            };
        }
        catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            const errorStack = error instanceof Error ? error.stack : undefined;
            return this.createErrorResponse(request.workflowId, requestId, 'EXECUTION_ERROR', errorMessage, startTime, errorStack);
        }
    }
    /**
     * Resolve a workflow from the registry
     */
    async resolveWorkflow(workflowId) {
        if (!this.registry) {
            return null;
        }
        return this.registry.getEndpoint(workflowId);
    }
    /**
     * Execute workflow with timeout support
     */
    async executeWithTimeout(filePath, params, options, timeout, abortSignal) {
        // Create a promise that rejects on timeout
        const timeoutPromise = new Promise((_, reject) => {
            const timer = setTimeout(() => {
                reject(new Error(`Workflow execution timed out after ${timeout}ms`));
            }, timeout);
            // Clear timeout if abort signal fires
            if (abortSignal) {
                abortSignal.addEventListener('abort', () => {
                    clearTimeout(timer);
                    reject(new Error('Workflow execution was cancelled'));
                });
            }
        });
        // Race between execution and timeout
        return Promise.race([executeWorkflowFromFile(filePath, params, options), timeoutPromise]);
    }
    /**
     * Create a standardized error response
     */
    createErrorResponse(workflowId, requestId, code, message, startTime, stack) {
        const error = {
            code,
            message,
        };
        // Only include stack in non-production environments
        if (stack && !this.production) {
            error.stack = stack;
        }
        return {
            success: false,
            workflowId,
            error,
            executionTime: Date.now() - startTime,
            requestId,
        };
    }
    /**
     * Validate a workflow request
     */
    validateRequest(request) {
        const errors = [];
        if (!request.workflowId) {
            errors.push({
                path: 'workflowId',
                message: 'Workflow ID is required',
            });
        }
        if (request.params && typeof request.params !== 'object') {
            errors.push({
                path: 'params',
                message: 'Params must be an object',
                expected: 'object',
                actual: typeof request.params,
            });
        }
        return {
            valid: errors.length === 0,
            errors: errors.length > 0 ? errors : undefined,
        };
    }
}
/**
 * Create a new executor instance
 */
export function createExecutor(options) {
    return new UnifiedWorkflowExecutor(options);
}
