// @synergenius/flow-weaver/dist/function-like.js
/**
 * FunctionLike interface
 *
 * Abstracts over FunctionDeclaration and arrow/function-expression variable
 * declarations so the JSDoc parser and annotation parser can handle both
 * syntactic forms uniformly.
 */
import { Node, SyntaxKind, VariableDeclarationKind, } from 'ts-morph';
function wrapFunctionDeclaration(fn) {
    return {
        getJsDocs: () => fn.getJsDocs(),
        getName: () => fn.getName(),
        isAsync: () => fn.isAsync(),
        getParameters: () => fn.getParameters(),
        getReturnType: () => fn.getReturnType(),
        getStartLineNumber: (inc) => fn.getStartLineNumber(inc),
        getText: (inc) => fn.getText(inc),
        getSourceFile: () => fn.getSourceFile(),
        getTypeResolutionNode: () => fn,
        isAmbient: () => !fn.hasBody(),
    };
}
/**
 * Scan a source file and return every function-like declaration that could
 * carry @flowWeaver annotations:
 *
 *  1. Classic `function foo() {}`         — FunctionDeclaration (existing)
 *  2. `const foo = () => {}`              — ArrowFunction assigned to variable
 *  3. `const foo = function() {}`         — FunctionExpression assigned to variable
 *
 * The returned list preserves source order.
 */
export function extractFunctionLikes(sourceFile) {
    const results = [];
    // 1. Regular function declarations — wrapped to satisfy getTypeResolutionNode()
    for (const fn of sourceFile.getFunctions()) {
        results.push(wrapFunctionDeclaration(fn));
    }
    // 2. Variable declarations with arrow/function expression initializers
    for (const varDecl of sourceFile.getVariableDeclarations()) {
        const initializer = varDecl.getInitializer();
        if (!initializer)
            continue;
        const kind = initializer.getKind();
        if (kind !== SyntaxKind.ArrowFunction && kind !== SyntaxKind.FunctionExpression)
            continue;
        const fnExpr = initializer;
        // Walk up: VariableDeclaration -> VariableDeclarationList -> VariableStatement
        const varDeclList = varDecl.getParent();
        if (!Node.isVariableDeclarationList(varDeclList))
            continue;
        const varStatement = varDeclList.getParent();
        if (!Node.isVariableStatement(varStatement))
            continue;
        // Determine declaration kind (const/let/var)
        const declKindEnum = varDeclList.getDeclarationKind();
        const declKind = declKindEnum === VariableDeclarationKind.Let
            ? 'let'
            : declKindEnum === VariableDeclarationKind.Var
                ? 'var'
                : 'const';
        const adapter = {
            getName: () => varDecl.getName(),
            getJsDocs: () => varStatement.getJsDocs(),
            isAsync: () => fnExpr.isAsync(),
            getParameters: () => fnExpr.getParameters(),
            getReturnType: () => fnExpr.getReturnType(),
            getStartLineNumber: () => varStatement.getStartLineNumber(false),
            getText: () => {
                // Return the variable statement text without JSDoc (consistent with FunctionDeclaration.getText()).
                // The parser code manually prepends JSDoc via getJsDocs().
                return varStatement.getText(false);
            },
            getSourceFile: () => sourceFile,
            getDeclarationKind: () => declKind,
            getTypeResolutionNode: () => varDecl,
        };
        results.push(adapter);
    }
    // Sort by source position to preserve declaration order
    results.sort((a, b) => a.getStartLineNumber(false) - b.getStartLineNumber(false));
    return results;
}
