// @synergenius/flow-weaver/dist/cli/commands/listen.js
import { io as socketIO } from 'socket.io-client';
import { logger } from '../utils/logger.js';
import { DEFAULT_SERVER_URL } from '../../defaults.js';
export async function listenCommand(options) {
    const serverUrl = options.server || DEFAULT_SERVER_URL;
    const ioFactory = options.ioFactory ?? socketIO;
    const output = options.output ?? process.stdout;
    const socket = ioFactory(`${serverUrl}/integrations`, {
        query: { clientType: 'cli' },
        transports: ['websocket', 'polling'],
        reconnection: true,
        reconnectionDelay: 1000,
        reconnectionAttempts: Infinity,
    });
    socket.on('connect', () => {
        logger.info(`Connected to ${serverUrl}/integrations`);
        socket.emit('integration:getContext', (ctx) => {
            if (ctx) {
                output.write(JSON.stringify({ event: 'integration:context', data: ctx }) + '\n');
            }
        });
    });
    socket.on('disconnect', (reason) => {
        logger.warn(`Disconnected: ${reason}`);
    });
    socket.onAny((event, data) => {
        if (event.startsWith('fw:') || event.startsWith('integration:')) {
            output.write(JSON.stringify({ event, data }) + '\n');
        }
    });
    process.on('SIGINT', () => {
        logger.info('Disconnecting...');
        socket.disconnect();
        process.exit(0);
    });
    // Keep alive until SIGINT
    await new Promise(() => { });
}
