// @synergenius/flow-weaver/dist/runtime/checkpoint.js
/**
 * Checkpoint serialization for crash recovery.
 *
 * Writes workflow state to disk after each node completes. If the process
 * crashes, the checkpoint file persists and can be used to resume execution
 * from the last completed node.
 *
 * Checkpoint files live in .fw-checkpoints/ next to the workflow file and
 * are auto-deleted after successful workflow completion.
 */
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
// ---------------------------------------------------------------------------
// Serialization helpers
// ---------------------------------------------------------------------------
function isUnserializableMarker(value) {
    return (typeof value === 'object' &&
        value !== null &&
        value.__fw_unserializable__ === true);
}
/**
 * Try to serialize a value. If it contains functions, invoke them first.
 * If serialization fails, return a marker.
 */
function serializeValue(key, value, unsafeNodes) {
    // Resolve function values (pull execution lazy evaluation)
    if (typeof value === 'function') {
        try {
            value = value();
        }
        catch {
            const parts = key.split(':');
            unsafeNodes.add(parts[0]);
            return {
                __fw_unserializable__: true,
                nodeId: parts[0],
                portName: parts[1] || 'unknown',
                reason: 'Function invocation failed',
            };
        }
    }
    // Handle Promises: can't serialize, mark as unsafe
    if (value instanceof Promise) {
        const parts = key.split(':');
        unsafeNodes.add(parts[0]);
        return {
            __fw_unserializable__: true,
            nodeId: parts[0],
            portName: parts[1] || 'unknown',
            reason: 'Promise value',
        };
    }
    // Test serialization
    try {
        JSON.stringify(value);
        return value;
    }
    catch {
        const parts = key.split(':');
        unsafeNodes.add(parts[0]);
        return {
            __fw_unserializable__: true,
            nodeId: parts[0],
            portName: parts[1] || 'unknown',
            reason: 'Not JSON-serializable',
        };
    }
}
/**
 * Compute SHA-256 hash of a file's contents.
 */
function hashFile(filePath) {
    const content = fs.readFileSync(filePath, 'utf8');
    return crypto.createHash('sha256').update(content).digest('hex');
}
// ---------------------------------------------------------------------------
// CheckpointWriter
// ---------------------------------------------------------------------------
export class CheckpointWriter {
    dir;
    filePath;
    workflowName;
    runId;
    params;
    workflowHash;
    checkpointPath;
    writeLock = Promise.resolve();
    constructor(workflowFilePath, workflowName, runId, params = {}) {
        this.filePath = path.resolve(workflowFilePath);
        this.workflowName = workflowName;
        this.runId = runId;
        this.params = params;
        this.dir = path.join(path.dirname(this.filePath), '.fw-checkpoints');
        this.checkpointPath = path.join(this.dir, `${workflowName}-${runId}.json`);
        this.workflowHash = hashFile(this.filePath);
    }
    /**
     * Write a checkpoint after a node completes. Uses a write lock so
     * concurrent calls from parallel nodes are serialized.
     */
    async write(completedNodes, executionOrder, position, ctx) {
        // Serialize under a lock to prevent concurrent writes (parallel nodes)
        this.writeLock = this.writeLock.then(() => this._writeCheckpoint(completedNodes, executionOrder, position, ctx));
        await this.writeLock;
    }
    /** Clean up checkpoint file after successful completion */
    cleanup() {
        try {
            if (fs.existsSync(this.checkpointPath)) {
                fs.unlinkSync(this.checkpointPath);
            }
            // Remove directory if empty
            if (fs.existsSync(this.dir)) {
                const remaining = fs.readdirSync(this.dir);
                if (remaining.length === 0) {
                    fs.rmdirSync(this.dir);
                }
            }
        }
        catch {
            // Best-effort cleanup
        }
    }
    getCheckpointPath() {
        return this.checkpointPath;
    }
    _writeCheckpoint(completedNodes, executionOrder, position, ctx) {
        const serialized = ctx.serialize();
        const unsafeNodes = new Set();
        // Serialize variables, handling unserializable values
        const variables = {};
        for (const [key, value] of Object.entries(serialized.variables)) {
            variables[key] = serializeValue(key, value, unsafeNodes);
        }
        const data = {
            version: 1,
            workflowHash: this.workflowHash,
            workflowName: this.workflowName,
            filePath: this.filePath,
            params: this.params,
            timestamp: new Date().toISOString(),
            completedNodes: [...completedNodes],
            executionOrder: [...executionOrder],
            position,
            variables,
            executions: serialized.executions,
            executionCounter: serialized.executionCounter,
            nodeExecutionCounts: serialized.nodeExecutionCounts,
            unsafeNodes: [...unsafeNodes],
        };
        // Ensure directory exists
        if (!fs.existsSync(this.dir)) {
            fs.mkdirSync(this.dir, { recursive: true });
        }
        fs.writeFileSync(this.checkpointPath, JSON.stringify(data, null, 2), 'utf8');
    }
}
// ---------------------------------------------------------------------------
// Checkpoint reading and resume
// ---------------------------------------------------------------------------
/**
 * Load a checkpoint file and validate it against the current workflow.
 * Returns the checkpoint data and a list of nodes that need to be re-run
 * (because their outputs weren't serializable).
 */
export function loadCheckpoint(checkpointPath, workflowFilePath) {
    const raw = fs.readFileSync(checkpointPath, 'utf8');
    const data = JSON.parse(raw);
    if (data.version !== 1) {
        throw new Error(`Unsupported checkpoint version: ${data.version}`);
    }
    // Check if the workflow has changed since the checkpoint was written
    let stale = false;
    if (workflowFilePath) {
        const currentHash = hashFile(path.resolve(workflowFilePath));
        stale = currentHash !== data.workflowHash;
    }
    // Determine which nodes can be skipped (all outputs serialized) vs
    // which need re-running (any output was unserializable)
    const unsafeSet = new Set(data.unsafeNodes);
    const rerunNodes = [];
    const skipNodes = new Map();
    // Walk execution order up to the checkpoint position.
    // Once we hit an unsafe node, everything from that point forward re-runs.
    let hitUnsafe = false;
    for (const nodeId of data.completedNodes) {
        if (hitUnsafe || unsafeSet.has(nodeId)) {
            hitUnsafe = true;
            rerunNodes.push(nodeId);
            continue;
        }
        // Collect this node's outputs from the variables map
        const nodeOutputs = {};
        const prefix = `${nodeId}:`;
        for (const [key, value] of Object.entries(data.variables)) {
            if (key.startsWith(prefix) && !isUnserializableMarker(value)) {
                // Store with portName:executionIndex as key
                const rest = key.substring(prefix.length);
                nodeOutputs[rest] = value;
            }
        }
        skipNodes.set(nodeId, nodeOutputs);
    }
    return { data, stale, rerunNodes, skipNodes };
}
/**
 * Find the most recent checkpoint file for a workflow.
 */
export function findLatestCheckpoint(workflowFilePath, workflowName) {
    const dir = path.join(path.dirname(path.resolve(workflowFilePath)), '.fw-checkpoints');
    if (!fs.existsSync(dir))
        return null;
    const files = fs.readdirSync(dir)
        .filter((f) => f.endsWith('.json'))
        .filter((f) => !workflowName || f.startsWith(`${workflowName}-`));
    if (files.length === 0)
        return null;
    // Sort by modification time, newest first
    const sorted = files
        .map((f) => ({ name: f, mtime: fs.statSync(path.join(dir, f)).mtimeMs }))
        .sort((a, b) => b.mtime - a.mtime);
    return path.join(dir, sorted[0].name);
}
