// @synergenius/flow-weaver/dist/runtime/debug-controller.js
/**
 * DebugController intercepts workflow execution at node boundaries,
 * enabling step-through debugging and checkpoint/resume.
 *
 * Injected via globalThis.__fw_debug_controller__ (same pattern as
 * __fw_debugger__ and __fw_agent_channel__). The generated code calls
 * beforeNode/afterNode at each node boundary; the controller decides
 * whether to skip, pause, checkpoint, or continue.
 */
// ---------------------------------------------------------------------------
// DebugController
// ---------------------------------------------------------------------------
export class DebugController {
    mode;
    breakpoints;
    completedNodes = [];
    completedSet = new Set();
    executionOrder = [];
    position = 0;
    lastCompletedNodeId = null;
    // Checkpoint
    checkpointEnabled;
    checkpointWriter;
    // Skip nodes (for resume from checkpoint)
    skipNodes;
    // Pause/resume channel (mirrors AgentChannel pattern)
    _gateResolve = null;
    _pauseResolve = null;
    _pausePromise;
    // Variable modification buffer: applied before next node runs
    pendingModifications = new Map();
    constructor(config = {}) {
        this.mode = config.debug ? 'step' : 'run';
        this.breakpoints = new Set(config.breakpoints ?? []);
        this.checkpointEnabled = config.checkpoint ?? false;
        this.checkpointWriter = config.checkpointWriter ?? null;
        this.executionOrder = config.executionOrder ?? [];
        this.skipNodes = config.skipNodes ?? new Map();
        this._pausePromise = this._createPausePromise();
    }
    /** Set the execution order (called by executor after compilation) */
    setExecutionOrder(order) {
        this.executionOrder = order;
    }
    // -----------------------------------------------------------------------
    // Node boundary hooks (called by generated code)
    // -----------------------------------------------------------------------
    /**
     * Called before a node executes.
     * Returns true if the node should execute, false to skip.
     */
    async beforeNode(nodeId, ctx) {
        // Apply any pending variable modifications
        this.applyPendingModifications(ctx);
        // If this node should be skipped (resume from checkpoint), restore its
        // outputs into the context and return false
        if (this.skipNodes.has(nodeId)) {
            const savedOutputs = this.skipNodes.get(nodeId);
            this.restoreNodeOutputs(nodeId, savedOutputs, ctx);
            this.completedNodes.push(nodeId);
            this.completedSet.add(nodeId);
            this.lastCompletedNodeId = nodeId;
            this.position++;
            return false;
        }
        // Check if we should pause here
        const shouldPause = this.mode === 'step' ||
            (this.mode === 'continueToBreakpoint' && this.breakpoints.has(nodeId));
        if (shouldPause) {
            const action = await this.pause(nodeId, 'before', ctx);
            if (action.type === 'abort') {
                throw new Error(`Debug session aborted at node "${nodeId}"`);
            }
            // Action may change mode for subsequent nodes
            this.applyAction(action);
        }
        return true;
    }
    /**
     * Called after a node completes successfully.
     */
    async afterNode(nodeId, ctx) {
        this.completedNodes.push(nodeId);
        this.completedSet.add(nodeId);
        this.lastCompletedNodeId = nodeId;
        this.position++;
        // Write checkpoint to disk
        if (this.checkpointEnabled && this.checkpointWriter) {
            await this.checkpointWriter.write(this.completedNodes, this.executionOrder, this.position, ctx);
        }
        // Pause after node in step mode
        if (this.mode === 'step') {
            const action = await this.pause(nodeId, 'after', ctx);
            if (action.type === 'abort') {
                throw new Error(`Debug session aborted after node "${nodeId}"`);
            }
            this.applyAction(action);
        }
    }
    // -----------------------------------------------------------------------
    // Pause/resume channel
    // -----------------------------------------------------------------------
    /**
     * Awaited by the executor to detect when the controller pauses.
     * Resolves with the current debug state.
     */
    onPause() {
        return this._pausePromise;
    }
    /**
     * Called by MCP tools or CLI to resume execution.
     */
    resume(action) {
        if (action.type !== 'abort') {
            this.applyAction(action);
        }
        this._gateResolve?.(action);
        this._gateResolve = null;
        this._pausePromise = this._createPausePromise();
    }
    // -----------------------------------------------------------------------
    // Variable modification
    // -----------------------------------------------------------------------
    /**
     * Queue a variable modification. Applied before the next node runs.
     * Key format: "nodeId:portName:executionIndex"
     */
    setVariable(key, value) {
        this.pendingModifications.set(key, value);
    }
    // -----------------------------------------------------------------------
    // Breakpoints
    // -----------------------------------------------------------------------
    addBreakpoint(nodeId) {
        this.breakpoints.add(nodeId);
    }
    removeBreakpoint(nodeId) {
        this.breakpoints.delete(nodeId);
    }
    getBreakpoints() {
        return [...this.breakpoints];
    }
    // -----------------------------------------------------------------------
    // State inspection
    // -----------------------------------------------------------------------
    /** Build the current debug state for external consumers */
    buildState(nodeId, phase, ctx) {
        const variables = this.extractVariables(ctx);
        const currentNodeOutputs = this.lastCompletedNodeId
            ? this.extractNodeOutputs(this.lastCompletedNodeId, variables)
            : undefined;
        return {
            currentNodeId: nodeId,
            phase,
            completedNodes: [...this.completedNodes],
            executionOrder: [...this.executionOrder],
            position: this.position,
            variables,
            currentNodeOutputs,
            breakpoints: [...this.breakpoints],
        };
    }
    /** Get completed nodes list */
    getCompletedNodes() {
        return [...this.completedNodes];
    }
    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------
    async pause(nodeId, phase, ctx) {
        const state = this.buildState(nodeId, phase, ctx);
        // Signal the executor that we're paused
        this._pauseResolve?.(state);
        // Suspend on a gate Promise until resume() is called
        return new Promise((resolve) => {
            this._gateResolve = (action) => resolve(action);
        });
    }
    applyAction(action) {
        switch (action.type) {
            case 'step':
                this.mode = 'step';
                break;
            case 'continue':
                this.mode = 'continue';
                break;
            case 'continueToBreakpoint':
                this.mode = 'continueToBreakpoint';
                break;
            // 'abort' is handled by the caller (throws)
        }
    }
    applyPendingModifications(ctx) {
        if (this.pendingModifications.size === 0)
            return;
        for (const [key, value] of this.pendingModifications) {
            // Key format: "nodeId:portName:executionIndex"
            const parts = key.split(':');
            if (parts.length >= 3) {
                const address = {
                    id: parts[0],
                    portName: parts[1],
                    executionIndex: parseInt(parts[2], 10),
                };
                ctx.setVariable(address, value);
            }
        }
        this.pendingModifications.clear();
    }
    restoreNodeOutputs(nodeId, outputs, ctx) {
        // outputs is keyed by "portName:executionIndex" -> value
        // Don't call ctx.addExecution here: the generated code's else block handles
        // execution registration so local variables (e.g. dIdx) are set correctly.
        for (const [portKey, value] of Object.entries(outputs)) {
            const colonIdx = portKey.lastIndexOf(':');
            if (colonIdx === -1)
                continue;
            const portName = portKey.substring(0, colonIdx);
            const executionIndex = parseInt(portKey.substring(colonIdx + 1), 10);
            ctx.setVariable({ id: nodeId, portName, executionIndex }, value);
        }
    }
    extractVariables(ctx) {
        const serialized = ctx.serialize();
        return serialized.variables;
    }
    extractNodeOutputs(nodeId, allVariables) {
        const outputs = {};
        const prefix = `${nodeId}:`;
        for (const [key, value] of Object.entries(allVariables)) {
            if (key.startsWith(prefix)) {
                // Extract portName from key "nodeId:portName:executionIndex"
                const rest = key.substring(prefix.length);
                const colonIdx = rest.lastIndexOf(':');
                const portName = colonIdx >= 0 ? rest.substring(0, colonIdx) : rest;
                outputs[portName] = value;
            }
        }
        return outputs;
    }
    _createPausePromise() {
        return new Promise((resolve) => {
            this._pauseResolve = resolve;
        });
    }
}
