// @synergenius/flow-weaver/dist/runtime/function-registry.js
/**
 * Function Registry for Flow Weaver
 *
 * Provides a global registry of functions that can be referenced by string IDs
 * in external HTTP calls while allowing direct function passing in internal calls.
 */
export class FunctionRegistry {
    functions = new Map();
    /**
     * Register a function with the registry
     */
    register(config) {
        if (this.functions.has(config.id)) {
            throw new Error(`Function with ID '${config.id}' is already registered`);
        }
        this.functions.set(config.id, config);
    }
    /**
     * Get a function by ID
     */
    get(id) {
        const registered = this.functions.get(id);
        return registered?.fn;
    }
    /**
     * Get function metadata without the function itself
     */
    getMetadata(id) {
        const registered = this.functions.get(id);
        if (!registered)
            return undefined;
        const { fn: _, ...metadata } = registered;
        return metadata;
    }
    /**
     * List all registered functions, optionally filtered by category
     */
    list(category) {
        const all = Array.from(this.functions.values());
        const filtered = category ? all.filter((f) => f.category === category) : all;
        return filtered.map(({ fn: _, ...metadata }) => metadata);
    }
    /**
     * Check if a function is registered
     */
    has(id) {
        return this.functions.has(id);
    }
    /**
     * Clear all registered functions (useful for testing)
     */
    clear() {
        this.functions.clear();
    }
    /**
     * Get the count of registered functions
     */
    get size() {
        return this.functions.size;
    }
}
/**
 * Global function registry singleton
 */
export const functionRegistry = new FunctionRegistry();
