// @synergenius/flow-weaver/dist/runtime/parameter-resolver.js
/**
 * Parameter Resolver for Flow Weaver
 *
 * Resolves function-like parameters to executable functions.
 * Supports:
 * - Direct functions (passed through)
 * - Registry IDs (looked up in function registry)
 * - Partial application (registry ID with pre-bound arguments)
 */
import { functionRegistry } from './function-registry.js';
/**
 * Resolve a function-like parameter to an executable function
 *
 * @param param - The function-like parameter to resolve
 * @returns The resolved function with metadata
 * @throws Error if registry ID is not found
 *
 * @example
 * // Direct function
 * const resolved = resolveFunction((x: number) => x * 2);
 * resolved.fn(5); // 10
 *
 * @example
 * // Registry ID
 * const resolved = resolveFunction('string:uppercase');
 * resolved.fn('hello'); // 'HELLO'
 *
 * @example
 * // Partial application
 * const resolved = resolveFunction({
 *   registryId: 'math:clamp',
 *   partialArgs: { min: 0, max: 100 }
 * });
 * resolved.fn({ value: 150 }); // 100
 */
export function resolveFunction(param) {
    // Direct function - pass through
    if (typeof param === 'function') {
        return {
            fn: param,
            source: 'direct',
        };
    }
    // String registry ID
    if (typeof param === 'string') {
        const fn = functionRegistry.get(param);
        if (!fn) {
            throw new Error(`Function '${param}' not found in registry`);
        }
        return {
            fn,
            source: 'registry',
            registryId: param,
        };
    }
    // Object with registry ID and optional partial args
    if (typeof param === 'object' && param !== null && 'registryId' in param) {
        const fn = functionRegistry.get(param.registryId);
        if (!fn) {
            throw new Error(`Function '${param.registryId}' not found in registry`);
        }
        // If partial args provided, create wrapper that merges them
        if (param.partialArgs) {
            const partialArgs = param.partialArgs;
            const wrappedFn = (input) => {
                // Merge partial args with input (input takes precedence for conflicting keys)
                const mergedInput = typeof input === 'object' && input !== null
                    ? { ...partialArgs, ...input }
                    : input;
                return fn(mergedInput);
            };
            return {
                fn: wrappedFn,
                source: 'registry',
                registryId: param.registryId,
            };
        }
        return {
            fn,
            source: 'registry',
            registryId: param.registryId,
        };
    }
    throw new Error(`Invalid function parameter: ${JSON.stringify(param)}`);
}
/**
 * Check if a value is a function-like parameter
 */
export function isFunctionLike(value) {
    if (typeof value === 'function')
        return true;
    if (typeof value === 'string')
        return true;
    if (typeof value === 'object' &&
        value !== null &&
        'registryId' in value &&
        typeof value.registryId === 'string') {
        return true;
    }
    return false;
}
/**
 * Try to resolve a function, returning undefined if not found
 */
export function tryResolveFunction(param) {
    try {
        return resolveFunction(param);
    }
    catch {
        return undefined;
    }
}
