// @synergenius/flow-weaver/dist/cli/tunnel/handlers/ast-ops.js
/**
 * AST read-operation handlers for the tunnel CLI.
 * Ported from flow-weaver-platform/src/routes/studio-rpc.ts + src/services/ast-helpers.ts.
 */
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { parser, resolveNpmNodeTypes } from '../../../parser.js';
import { validateWorkflow } from '../../../api/validate.js';
import { resolvePath, toVirtualPath } from '../path-resolver.js';
// ---------------------------------------------------------------------------
// Helpers — ported from studio-rpc.ts lines 98-186
// ---------------------------------------------------------------------------
function ensureASTDefaults(ast) {
    return {
        ...ast,
        instances: ast.instances ?? [],
        connections: ast.connections ?? [],
        nodeTypes: ast.nodeTypes ?? [],
    };
}
function virtualizeASTPaths(ast, wsPath) {
    const result = { ...ast };
    // 1. sourceFile
    if (typeof result.sourceFile === 'string' && result.sourceFile.startsWith(wsPath)) {
        result.sourceFile = toVirtualPath(wsPath, result.sourceFile);
    }
    // 2. nodeTypes[].sourceLocation.file and nodeTypes[].path
    if (Array.isArray(result.nodeTypes)) {
        result.nodeTypes = result.nodeTypes.map((nt) => {
            const copy = { ...nt };
            if (copy.sourceLocation?.file?.startsWith(wsPath)) {
                copy.sourceLocation = { ...copy.sourceLocation, file: toVirtualPath(wsPath, copy.sourceLocation.file) };
            }
            if (typeof copy.path === 'string' && copy.path.startsWith(wsPath)) {
                copy.path = toVirtualPath(wsPath, copy.path);
            }
            return copy;
        });
    }
    // 3. instances[].sourceLocation.file
    if (Array.isArray(result.instances)) {
        result.instances = result.instances.map((inst) => {
            if (inst.sourceLocation?.file?.startsWith(wsPath)) {
                return {
                    ...inst,
                    sourceLocation: { ...inst.sourceLocation, file: toVirtualPath(wsPath, inst.sourceLocation.file) },
                };
            }
            return inst;
        });
    }
    return result;
}
function prepareMutationResult(ast, wsPath) {
    return virtualizeASTPaths(ensureASTDefaults(ast), wsPath);
}
function getWorkflowName(params) {
    return (params.functionName || params.workflowName || params.exportName);
}
// ---------------------------------------------------------------------------
// Core AST operations — ported from ast-helpers.ts
// ---------------------------------------------------------------------------
async function loadWorkflowAST(filePath, functionName) {
    const parsed = parser.parse(filePath);
    const workflows = parsed.workflows || [];
    if (workflows.length === 0) {
        throw new Error(`No workflows found in ${filePath}`);
    }
    const target = functionName
        ? workflows.find((w) => w.functionName === functionName)
        : workflows[0];
    if (!target) {
        throw new Error(`Workflow "${functionName}" not found in ${filePath}`);
    }
    return resolveNpmNodeTypes(target, path.dirname(filePath));
}
async function loadAllWorkflowsAST(wsPath) {
    const entries = await fs.readdir(wsPath, { withFileTypes: true });
    const results = [];
    for (const entry of entries) {
        if (!entry.isFile() || !entry.name.endsWith('.ts'))
            continue;
        if (entry.name === 'tsconfig.json' || entry.name === 'package.json')
            continue;
        const fullPath = path.join(wsPath, entry.name);
        try {
            const parsed = parser.parse(fullPath);
            const workflows = parsed.workflows || [];
            for (const wf of workflows) {
                const resolved = await resolveNpmNodeTypes(wf, wsPath);
                results.push({ filePath: fullPath, ast: resolved });
            }
        }
        catch {
            // Skip files that fail to parse
        }
    }
    return results;
}
function parseWorkflowFromContent(content) {
    const parsed = parser.parseFromString(content);
    const workflows = parsed.workflows || [];
    return workflows.map((wf) => ({
        name: wf.name || wf.functionName,
        ast: wf,
    }));
}
function getDiagnostics(source) {
    const parsed = parser.parseFromString(source);
    if ((parsed.errors && parsed.errors.length > 0) || !parsed.workflows?.length) {
        return {
            valid: false,
            errors: (parsed.errors || []).map((e) => ({ message: typeof e === 'string' ? e : e.message || String(e) })),
            warnings: [],
        };
    }
    const allErrors = [];
    const allWarnings = [];
    for (const wf of parsed.workflows) {
        const result = validateWorkflow(wf);
        if (result.errors)
            allErrors.push(...result.errors);
        if (result.warnings)
            allWarnings.push(...result.warnings);
    }
    return { valid: allErrors.length === 0, errors: allErrors, warnings: allWarnings };
}
async function extractAllNodeTypes(wsPath) {
    const all = await loadAllWorkflowsAST(wsPath);
    const typeMap = new Map();
    for (const { ast } of all) {
        for (const nt of ast.nodeTypes || []) {
            if (!typeMap.has(nt.name)) {
                typeMap.set(nt.name, nt);
            }
        }
    }
    return Array.from(typeMap.values());
}
// ---------------------------------------------------------------------------
// Exported handlers
// ---------------------------------------------------------------------------
export { prepareMutationResult, getWorkflowName };
export const astOpsHandlers = {
    loadWorkflowAST: async (params, ctx) => {
        const filePath = params.filePath;
        if (!filePath)
            throw new Error('filePath is required');
        const functionName = getWorkflowName(params);
        const resolved = resolvePath(ctx.workspaceRoot, filePath);
        const ast = await loadWorkflowAST(resolved, functionName);
        return virtualizeASTPaths(ensureASTDefaults(ast), ctx.workspaceRoot);
    },
    loadAllWorkflowsAST: async (_params, ctx) => {
        const all = await loadAllWorkflowsAST(ctx.workspaceRoot);
        return all.map(({ filePath, ast }) => ({
            filePath: toVirtualPath(ctx.workspaceRoot, filePath),
            ast: virtualizeASTPaths(ensureASTDefaults(ast), ctx.workspaceRoot),
        }));
    },
    parseWorkflowASTFromContent: async (params) => {
        const content = params.content;
        if (!content)
            throw new Error('content is required');
        const functionName = getWorkflowName(params);
        const results = parseWorkflowFromContent(content);
        if (functionName) {
            const match = results.find((r) => r.name === functionName || r.ast?.functionName === functionName);
            return match ? ensureASTDefaults(match.ast) : null;
        }
        return results.length > 0 ? ensureASTDefaults(results[0].ast) : null;
    },
    getAvailableWorkflowsInFile: async (params, ctx) => {
        const filePath = params.filePath;
        if (!filePath)
            return { availableWorkflows: [] };
        try {
            const resolved = resolvePath(ctx.workspaceRoot, filePath);
            const source = await fs.readFile(resolved, 'utf-8');
            const results = parseWorkflowFromContent(source);
            return {
                availableWorkflows: results.map((w) => ({
                    name: w.ast?.name || w.name,
                    functionName: w.ast?.functionName || w.name,
                    isExported: true,
                })),
            };
        }
        catch {
            return { availableWorkflows: [] };
        }
    },
    getDiagnostics: async (params, ctx) => {
        // Source resolution priority: openFiles > content > file on disk
        let source;
        const openFiles = params.openFiles;
        const filePath = params.filePath;
        if (openFiles && filePath && openFiles[filePath]) {
            source = openFiles[filePath];
        }
        else if (typeof params.content === 'string') {
            source = params.content;
        }
        else if (filePath) {
            const resolved = resolvePath(ctx.workspaceRoot, filePath);
            source = await fs.readFile(resolved, 'utf-8');
        }
        if (!source)
            throw new Error('No source provided for diagnostics');
        const { errors, warnings } = getDiagnostics(source);
        // Transform to flat array for FileEditorPane
        const result = [];
        for (const e of errors) {
            result.push({
                severity: 'error',
                message: e.message || String(e),
                start: e.location ?? { line: 1, column: 0 },
            });
        }
        for (const w of warnings) {
            result.push({
                severity: 'warning',
                message: w.message || String(w),
                start: w.location ?? { line: 1, column: 0 },
            });
        }
        return result;
    },
    getNodeTypes: async (_params, ctx) => {
        const types = await extractAllNodeTypes(ctx.workspaceRoot);
        return types.map((t) => ({ name: t.name, label: t.name, nodeType: t }));
    },
    getNodeTypesBatch: async (params, ctx) => {
        const cursor = params.cursor || '0';
        const limit = params.limit || 50;
        const offset = parseInt(cursor, 10) || 0;
        const types = await extractAllNodeTypes(ctx.workspaceRoot);
        const all = types.map((t) => ({ name: t.name, label: t.name, nodeType: t }));
        const page = all.slice(offset, offset + limit);
        const nextCursor = offset + limit < all.length ? String(offset + limit) : null;
        return { types: page, cursor: nextCursor };
    },
    searchNodeTypes: async (params, ctx) => {
        const query = (params.query || '').toLowerCase();
        const types = await extractAllNodeTypes(ctx.workspaceRoot);
        const all = types.map((t) => ({ name: t.name, label: t.name, nodeType: t }));
        if (!query)
            return all;
        return all.filter((t) => t.name.toLowerCase().includes(query) || t.label.toLowerCase().includes(query));
    },
};
