// @synergenius/flow-weaver/dist/cli/tunnel/handlers/execution.js
/**
 * Execution handlers for the tunnel CLI.
 * executeFile, compileFile, generateDiagram
 */
import * as fs from 'node:fs/promises';
import { compileWorkflow } from '../../../api/compile.js';
import { resolvePath } from '../path-resolver.js';
let sourceToSVG;
async function loadDiagram() {
    if (sourceToSVG)
        return sourceToSVG;
    try {
        const mod = await import('../../../diagram/index.js');
        sourceToSVG = mod.sourceToSVG;
        return sourceToSVG;
    }
    catch {
        return undefined;
    }
}
export const executionHandlers = {
    executeFile: async (params, ctx) => {
        const filePath = params.filePath;
        if (!filePath)
            throw new Error('filePath is required');
        const resolved = resolvePath(ctx.workspaceRoot, filePath);
        const inputData = (params.inputData || params.input || {});
        try {
            // Dynamic import to avoid bundling issues — uses the MCP workflow executor
            const { executeWorkflowFromFile } = await import('../../../mcp/workflow-executor.js');
            const result = await executeWorkflowFromFile(resolved, inputData);
            return result;
        }
        catch (err) {
            return {
                success: false,
                error: err instanceof Error ? err.message : String(err),
            };
        }
    },
    compileFile: async (params, ctx) => {
        const filePath = params.filePath;
        if (!filePath)
            throw new Error('filePath is required');
        const resolved = resolvePath(ctx.workspaceRoot, filePath);
        try {
            const result = await compileWorkflow(resolved);
            return result;
        }
        catch (err) {
            return {
                success: false,
                error: err instanceof Error ? err.message : String(err),
            };
        }
    },
    generateDiagram: async (params, ctx) => {
        const filePath = params.filePath;
        const content = params.content;
        let source;
        if (content) {
            source = content;
        }
        else if (filePath) {
            const resolved = resolvePath(ctx.workspaceRoot, filePath);
            source = await fs.readFile(resolved, 'utf-8');
        }
        else {
            throw new Error('filePath or content is required');
        }
        const render = await loadDiagram();
        if (!render) {
            return { success: false, error: 'Diagram module not available' };
        }
        try {
            const svg = render(source, {
                workflowName: params.workflowName,
            });
            return { success: true, svg };
        }
        catch (err) {
            return {
                success: false,
                error: err instanceof Error ? err.message : String(err),
            };
        }
    },
};
