// @synergenius/flow-weaver/dist/cli/tunnel/handlers/mutations.js
/**
 * Workflow mutation handlers for the tunnel CLI.
 * Ported from flow-weaver-platform/src/services/ast-helpers.ts mutation wrappers.
 */
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { parser, resolveNpmNodeTypes } from '../../../parser.js';
import { generateInPlace } from '../../../api/generate-in-place.js';
import { addNode, removeNode, updateNode, renameNode, setNodePosition, setNodeMinimized, setNodeSize, } from '../../../api/manipulation/nodes.js';
import { addConnection, removeConnection } from '../../../api/manipulation/connections.js';
import { addNodeType, removeNodeType, renameNodeType, updateNodeType, } from '../../../api/manipulation/node-types.js';
import { setStartExitPorts, setInstancePortConfigs } from '../../../api/manipulation/ports.js';
import { resolvePath } from '../path-resolver.js';
import { withFileLock } from '../file-lock.js';
import { prepareMutationResult, getWorkflowName } from './ast-ops.js';
// ---------------------------------------------------------------------------
// Core mutation engine — ported from ast-helpers.ts:95-159
// ---------------------------------------------------------------------------
async function mutateWorkflowFile({ filePath, functionName, mutator, }) {
    return withFileLock(filePath, async () => {
        const sourceCode = await fs.readFile(filePath, 'utf-8');
        const parsed = parser.parse(filePath);
        const workflows = parsed.workflows || [];
        if (workflows.length === 0) {
            throw new Error(`No workflows found in ${filePath}`);
        }
        const targetIndex = functionName
            ? workflows.findIndex((w) => w.functionName === functionName)
            : 0;
        if (targetIndex < 0) {
            throw new Error(`Workflow "${functionName}" not found in ${filePath}`);
        }
        const original = workflows[targetIndex];
        const updated = mutator(original);
        // Preserve importSource for node types through mutation
        const importSourceMap = new Map();
        for (const nt of original.nodeTypes || []) {
            if (nt.importSource) {
                importSourceMap.set(nt.name, nt.importSource);
            }
        }
        const updatedNodeTypes = (updated.nodeTypes || []).map((nt) => {
            if (!nt.importSource && importSourceMap.has(nt.name)) {
                return { ...nt, importSource: importSourceMap.get(nt.name) };
            }
            return nt;
        });
        // Re-append node types that existed in original but are missing after mutation
        const updatedTypeNames = new Set(updatedNodeTypes.map((nt) => nt.name));
        const missingTypes = [];
        for (const nt of original.nodeTypes || []) {
            if (!updatedTypeNames.has(nt.name)) {
                missingTypes.push(nt);
            }
        }
        const workflowForGeneration = {
            ...updated,
            nodeTypes: [...updatedNodeTypes, ...missingTypes],
        };
        const result = generateInPlace(sourceCode, workflowForGeneration);
        await fs.writeFile(filePath, result.code, 'utf-8');
        return resolveNpmNodeTypes(updated, path.dirname(filePath));
    });
}
// ---------------------------------------------------------------------------
// Handler factory — reduces boilerplate for simple mutations
// ---------------------------------------------------------------------------
function makeMutationHandler(extractMutator) {
    return async (params, ctx) => {
        const filePath = params.filePath;
        if (!filePath)
            throw new Error('filePath is required');
        const functionName = getWorkflowName(params);
        const resolved = resolvePath(ctx.workspaceRoot, filePath);
        const result = await mutateWorkflowFile({
            filePath: resolved,
            functionName,
            mutator: extractMutator(params),
        });
        return prepareMutationResult(result, ctx.workspaceRoot);
    };
}
// ---------------------------------------------------------------------------
// Exported handlers
// ---------------------------------------------------------------------------
export const mutationHandlers = {
    addNode: makeMutationHandler((params) => {
        let node = params.node;
        if (!node) {
            const nodeType = params.nodeType;
            const nodeName = params.nodeName;
            const position = params.position;
            node = { type: 'NodeInstance', id: nodeName, nodeType, ...(position && { position }) };
        }
        return (wf) => addNode(wf, node);
    }),
    removeNode: makeMutationHandler((params) => {
        const nodeId = (params.nodeName || params.nodeId);
        if (!nodeId)
            throw new Error('nodeId is required');
        return (wf) => removeNode(wf, nodeId, { removeConnections: true });
    }),
    updateNode: makeMutationHandler((params) => {
        const nodeId = (params.nodeId || params.nodeName);
        const updates = params.updates;
        if (!nodeId)
            throw new Error('nodeId is required');
        return (wf) => updateNode(wf, nodeId, updates);
    }),
    renameNode: makeMutationHandler((params) => {
        const oldId = params.oldId;
        const newId = params.newId;
        if (!oldId || !newId)
            throw new Error('oldId and newId are required');
        return (wf) => renameNode(wf, oldId, newId);
    }),
    setNodePosition: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const x = params.x;
        const y = params.y;
        return (wf) => setNodePosition(wf, nodeId, x, y);
    }),
    setNodePositions: makeMutationHandler((params) => {
        const positions = params.positions;
        if (!positions)
            throw new Error('positions are required');
        return (wf) => {
            let result = wf;
            for (const { nodeId, x, y } of positions) {
                result = setNodePosition(result, nodeId, x, y);
            }
            return result;
        };
    }),
    setNodeMinimized: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const minimized = params.minimized;
        return (wf) => setNodeMinimized(wf, nodeId, minimized);
    }),
    setNodeSize: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const width = params.width;
        const height = params.height;
        return (wf) => setNodeSize(wf, nodeId, width, height);
    }),
    setNodeLabel: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const label = params.label;
        return (wf) => updateNode(wf, nodeId, { label });
    }),
    addNodes: makeMutationHandler((params) => {
        const nodes = params.nodes;
        if (!nodes)
            throw new Error('nodes are required');
        return (wf) => {
            let result = wf;
            for (const node of nodes) {
                result = addNode(result, node);
            }
            return result;
        };
    }),
    removeNodes: makeMutationHandler((params) => {
        const nodeIds = params.nodeIds;
        if (!nodeIds)
            throw new Error('nodeIds are required');
        return (wf) => {
            let result = wf;
            for (const nodeId of nodeIds) {
                result = removeNode(result, nodeId, { removeConnections: true });
            }
            return result;
        };
    }),
    addConnection: makeMutationHandler((params) => {
        const from = params.from || { node: params.fromNode, port: params.fromPort };
        const to = params.to || { node: params.toNode, port: params.toPort };
        return (wf) => addConnection(wf, from, to);
    }),
    removeConnection: makeMutationHandler((params) => {
        const from = params.from || { node: params.fromNode, port: params.fromPort };
        const to = params.to || { node: params.toNode, port: params.toPort };
        return (wf) => removeConnection(wf, from, to);
    }),
    setConnections: makeMutationHandler((params) => {
        const connections = params.connections;
        return (wf) => ({ ...wf, connections });
    }),
    addConnections: makeMutationHandler((params) => {
        const connections = params.connections;
        return (wf) => ({ ...wf, connections: [...wf.connections, ...connections] });
    }),
    updateNodePortConfig: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const portName = params.portName;
        const portConfig = params.portConfig;
        return (wf) => {
            const node = wf.instances?.find((n) => n.id === nodeId);
            if (!node)
                throw new Error(`Node "${nodeId}" not found`);
            const existing = node.config?.portConfigs || [];
            const idx = existing.findIndex((pc) => pc.portName === portName);
            const updated = [...existing];
            if (idx >= 0) {
                updated[idx] = { ...updated[idx], ...portConfig };
            }
            else {
                updated.push({ portName, ...portConfig });
            }
            return updateNode(wf, nodeId, { config: { ...node.config, portConfigs: updated } });
        };
    }),
    resetNodePortConfig: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const portName = params.portName;
        return (wf) => {
            const node = wf.instances?.find((n) => n.id === nodeId);
            if (!node)
                throw new Error(`Node "${nodeId}" not found`);
            const existing = node.config?.portConfigs || [];
            const filtered = existing.filter((pc) => pc.portName !== portName);
            return updateNode(wf, nodeId, { config: { ...node.config, portConfigs: filtered } });
        };
    }),
    updateInstancePortConfigs: makeMutationHandler((params) => {
        const instanceId = params.instanceId;
        const portConfigs = params.portConfigs;
        return (wf) => setInstancePortConfigs(wf, instanceId, portConfigs);
    }),
    updateWorkflowPorts: makeMutationHandler((params) => {
        const nodeType = params.nodeType;
        const ports = params.ports;
        return (wf) => setStartExitPorts(wf, nodeType, ports);
    }),
    addNodeType: makeMutationHandler((params) => {
        const nodeType = params.nodeType;
        return (wf) => addNodeType(wf, nodeType);
    }),
    removeNodeType: makeMutationHandler((params) => {
        const typeName = params.typeName;
        if (!typeName)
            throw new Error('typeName is required');
        return (wf) => removeNodeType(wf, typeName);
    }),
    renameNodeType: makeMutationHandler((params) => {
        const oldTypeName = params.oldTypeName;
        const newTypeName = params.newTypeName;
        return (wf) => renameNodeType(wf, oldTypeName, newTypeName);
    }),
    updateNodeType: makeMutationHandler((params) => {
        const typeName = params.typeName;
        const updates = params.updates;
        return (wf) => updateNodeType(wf, typeName, updates);
    }),
    setNodeTypes: makeMutationHandler((params) => {
        const nodeTypes = params.nodeTypes;
        return (wf) => ({ ...wf, nodeTypes });
    }),
    saveWorkflowState: makeMutationHandler((params) => {
        const workflow = params.workflow;
        return () => workflow;
    }),
    setNodeParent: makeMutationHandler((params) => {
        const nodeId = params.nodeId;
        const parentId = params.parentId;
        return (wf) => updateNode(wf, nodeId, { parent: parentId });
    }),
    setNodesParent: makeMutationHandler((params) => {
        const nodeIds = params.nodeIds;
        const parentId = params.parentId;
        return (wf) => {
            let result = wf;
            for (const nodeId of nodeIds) {
                result = updateNode(result, nodeId, { parent: parentId });
            }
            return result;
        };
    }),
    setCurrentWorkflow: async () => ({ success: true }),
    setWorkflowForceAsync: makeMutationHandler((params) => {
        const forceAsync = params.forceAsync;
        return (wf) => ({ ...wf, forceAsync });
    }),
};
