// @synergenius/flow-weaver/dist/jsdoc-port-sync/index.js
/**
 * @module jsdoc-port-sync
 *
 * Browser-compatible bidirectional sync between JSDoc annotations and TypeScript
 * signatures. Uses regex only - no ts-morph dependency.
 *
 * ## Dual Source of Truth
 *
 * | Aspect         | Source        | Notes |
 * |----------------|---------------|-------|
 * | Port EXISTENCE | **Signature** | Params + return fields define what ports exist |
 * | Port TYPES     | **Signature** | Types inferred from TypeScript signature |
 * | Port METADATA  | **JSDoc**     | Labels, order, placement, scope |
 *
 * ## Sync Flow
 *
 * ```
 * Keystroke → syncJSDocToSignature (preserves signature, adds from JSDoc)
 *
 * Ctrl+P → syncSignatureToJSDoc (regenerates JSDoc) → syncJSDocToSignature
 * ```
 *
 * ## Key Functions
 *
 * - {@link parsePortsFromFunctionText} - Extract ports from JSDoc
 * - {@link syncJSDocToSignature} - JSDoc → Signature (add params/return fields)
 * - {@link syncSignatureToJSDoc} - Signature → JSDoc (regenerate tags)
 * - {@link syncCodeRenames} - Position-based rename detection
 * - {@link computePortsDiff} / {@link applyPortsDiffToCode} - Diff-based updates
 */
// Constants (kept for JSDoc detection/validation - Chevrotain handles parsing)
export { JSDOC_BLOCK_REGEX, PORT_TAG_REGEX, ORPHAN_PORT_LINE_REGEX, SCOPE_TAG_REGEX, FUNC_DECL_START, ARROW_START, RETURN_OBJECT_REGEX, RESERVED_PARAMS, RESERVED_RETURN_FIELDS, isReservedStepPort, findBalancedClose, splitParams, } from "./constants.js";
// Chevrotain-based port parsing (replaces regex patterns)
export { parsePortLine, isValidPortLine } from "../chevrotain-parser/port-parser.js";
// Signature parsing
export { parseFunctionSignature, parseReturnFields, parseReturnBodyFieldsWithTypes, parseReturnTypeFields, parseReturnTypeFieldsWithTypes, parseInputTypeFields, tsTypeToPortType, portTypeToTsType, parseCallbackType, buildCallbackType, callbackHasAllPorts, } from "./signature-parser.js";
// Port parsing
export { hasScopes, getScopeNames, hasOrphanPortLines, getIncompletePortNames, parsePortsFromFunctionText, updatePortsInFunctionText, } from "./port-parser.js";
// Sync functions
export { syncSignatureToJSDoc, syncJSDocToSignature, } from "./sync.js";
// Rename functions
export { renamePortInCode, syncCodeRenames, } from "./rename.js";
// Diff functions
export { computePortsDiff, formatPortsInFunctionText, applyPortsDiffToCode, } from "./diff.js";
