// @synergenius/flow-weaver/dist/chevrotain-parser/index.js
/**
 * @module chevrotain-parser
 *
 * Chevrotain-based parsers for JSDoc annotations.
 * Provides structured parsing for @input, @output, @node, @connect, and @position.
 */
// Token exports (for advanced use cases)
export { JSDocLexer, allTokens, InputTag, OutputTag, NodeTag, ConnectTag, PositionTag, ScopeTag, MapTag, PathTag, FanOutTag, FanInTag, CoerceTag, TriggerTag, CancelOnTag, RetriesTag, TimeoutTag, ThrottleTag, } from './tokens.js';
// Port parser
export { parsePortLine, getPortGrammar } from './port-parser.js';
// Node parser
export { parseNodeLine, getNodeGrammar } from './node-parser.js';
// Connect parser
export { parseConnectLine, getConnectGrammar } from './connect-parser.js';
// Position parser
export { parsePositionLine, getPositionGrammar } from './position-parser.js';
// Scope parser
export { parseScopeLine, getScopeGrammar } from './scope-parser.js';
// Map parser
export { parseMapLine, getMapGrammar } from './map-parser.js';
// Path parser
export { parsePathLine, getPathGrammar } from './path-parser.js';
// Fan-out / Fan-in parser
export { parseFanOutLine, parseFanInLine } from './fan-parser.js';
// Coerce parser
export { parseCoerceLine } from './coerce-parser.js';
// Trigger/cancel/retries/timeout/throttle parser
export { parseTriggerLine, parseCancelOnLine, parseRetriesLine, parseTimeoutLine, parseThrottleLine } from './trigger-cancel-parser.js';
// Grammar diagram generation
export { generateGrammarDiagrams, getAllGrammars, serializedToEBNF } from './grammar-diagrams.js';
