// @synergenius/flow-weaver/dist/chevrotain-parser/scope-parser.js
/**
 * @module chevrotain-parser/scope-parser
 *
 * Parser for @scope declarations using Chevrotain.
 */
import { CstParser } from 'chevrotain';
import { JSDocLexer, ScopeTag, Identifier, Dot, LBracket, RBracket, Comma, allTokens, } from './tokens.js';
// =============================================================================
// Parser Definition
// =============================================================================
class ScopeParser extends CstParser {
    constructor() {
        super(allTokens);
        this.performSelfAnalysis();
    }
    // Entry rule for scope line: @scope scopeName [child1, child2]
    // or @scope container.scopeName [child1, child2]
    scopeLine = this.RULE('scopeLine', () => {
        this.CONSUME(ScopeTag);
        this.SUBRULE(this.scopeRef);
        this.CONSUME(LBracket);
        this.AT_LEAST_ONE_SEP({
            SEP: Comma,
            DEF: () => {
                this.CONSUME(Identifier, { LABEL: 'childId' });
            },
        });
        this.CONSUME(RBracket);
    });
    // scopeName or container.scopeName
    scopeRef = this.RULE('scopeRef', () => {
        this.CONSUME(Identifier, { LABEL: 'firstPart' });
        this.OPTION(() => {
            this.CONSUME(Dot);
            this.CONSUME2(Identifier, { LABEL: 'secondPart' });
        });
    });
}
// =============================================================================
// Parser Instance (singleton)
// =============================================================================
const parserInstance = new ScopeParser();
// =============================================================================
// CST Visitor
// =============================================================================
const BaseVisitor = parserInstance.getBaseCstVisitorConstructor();
class ScopeVisitor extends BaseVisitor {
    constructor() {
        super();
        this.validateVisitor();
    }
    scopeLine(ctx) {
        const scopeName = this.visit(ctx.scopeRef);
        const children = ctx.childId.map((token) => token.image);
        return { scopeName, children };
    }
    scopeRef(ctx) {
        const firstPart = ctx.firstPart[0].image;
        if (ctx.secondPart) {
            const secondPart = ctx.secondPart[0].image;
            return `${firstPart}.${secondPart}`;
        }
        return firstPart;
    }
}
const visitorInstance = new ScopeVisitor();
// =============================================================================
// Public API
// =============================================================================
/**
 * Parse a @scope line and return structured result.
 * Returns null if the line is not a scope declaration.
 */
export function parseScopeLine(input, warnings) {
    const lexResult = JSDocLexer.tokenize(input);
    if (lexResult.errors.length > 0) {
        return null;
    }
    if (lexResult.tokens.length === 0) {
        return null;
    }
    const firstToken = lexResult.tokens[0];
    if (firstToken.tokenType !== ScopeTag) {
        return null;
    }
    parserInstance.input = lexResult.tokens;
    const cst = parserInstance.scopeLine();
    if (parserInstance.errors.length > 0) {
        const firstError = parserInstance.errors[0];
        const truncatedInput = input.length > 60 ? input.substring(0, 60) + '...' : input;
        warnings.push(`Failed to parse scope line: "${truncatedInput}"\n` +
            `  Error: ${firstError.message}\n` +
            `  Expected format: @scope scopeName [child1, child2]`);
        return null;
    }
    return visitorInstance.visit(cst);
}
/**
 * Get serialized grammar for documentation/diagram generation.
 */
export function getScopeGrammar() {
    return parserInstance.getSerializedGastProductions();
}
